core/src/main/scala/kafka/log/Log.scala

```diff
@@ -310,6 +310,10 @@ class Log(@volatile var dir: File,
     // from scratch.
     if (!producerStateManager.isEmpty)
       throw new IllegalStateException("Producer state must be empty during log initialization")
+    // Pass the set of loaded segment base offsets to diff against the set of producer state snapshot
+    // files. Any orphaned files should be cleaned up.
+    val keepOffsets = logSegments.map(_.baseOffset)
+    producerStateManager.cleanupOrphanSnapshotFiles(keepOffsets)
```

Comment:
Just mentioning this as an observation. Unlike the index files, which have a lifecycle clearly tied to the segment, producer snapshots are a bit of a mix. They are created in `Log` when a segment is rolled, but deletion is the job of the segment. I'm partially wondering if it would be clearer to handle both operations in the same context.

Reply:
Hmm I might be misunderstanding @hachikuji but the current PR should delete producer state snapshot files along with segment deletion. I included this additional cleanup step during log opening because snapshot files are not atomically renamed as part of deletion like other files, which seems to allow for orphaned files if there is a crash during segment deletion. Are you suggesting that we handle snapshot file deletion like the deletion of index/segment files by adding the `deleted` suffix?

Reply:
What I was saying is more in line with Jun's comment below. It is more about managing the lifecycle of the snapshot consistently. Typically it's simpler if creation and deletion happen in the same context.